File: gang/week12/짝지어 제거하기.java
import java.util.Stack;

class Solution {
    public int solution(String S) {
        Stack<Character> stack = new Stack<>(); // 스택 사용

        for (int i = 0; i < S.length(); i++) {
            char c = S.charAt(i);
            if (!stack.isEmpty() && stack.peek() == c) {    // 스택이 비워져 있지 않고, 맨 위 문자가 현재 문자와 같으면
                stack.pop();    // 맨위 문자 제거
            } else {
                stack.push(c);  // 아니면 현재 문자 삽입
            }
        }
        return stack.isEmpty() ? 1 : 0;
    }
}

File: yongkeun/12week/짝지어제거하기.js
function solution(s) {
  //baabaa 라고 가정할 때
  //stack 에 b가 박히고, a가 들어옴.
  //stack 에 ba가 박히고, a가 들어옴.
  //stack 마지막 a와 들어온 a가 같으므로 stack 마지막을 제거
  let stack = [];
  for (let i = 0; i < s.length; i++) {
    if (stack[stack.length - 1] === s[i]) {
      stack.pop();
    } else {
      stack.push(s[i]);
    }
  }
  return stack.length === 0 ? 1 : 0;
}
